### ISSUE TITLE
About the volumetric Aggregator's SSL

### ISSUE THREAD

**Issue by Zhong1015** [serag-ai/SAMF#5]
Hello authors! SAMF is a great work !
Recently, as I find that in your paper you used a warm up training for the aggregator, I 'd like to know whether it is a must as the result in the ablation study is bad if we do not apply SSL warm up training for the aggregator. If so, whether there is a code or a pretrained weight for this part?

Thank you for your great efforts for this work !

**Comment by abd-hsn** [comment]
Hi @Zhong1015 
Regarding the warmup step for the aggregator, **it is not mandatory.** We included this step only as part of an ablation study to evaluate whether it improves model performance.
For reference, I’ll attach the code for the warmup step. However, please note that **this stage is optional**, and you should be able to reproduce the results reported in the paper without performing warmup training for the aggregator.

```python
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision.transforms import transforms as T
import nibabel as nib
import numpy as np
from torch.utils.data import Dataset
import matplotlib.pyplot as plt
from PIL import Image
import random
from tqdm import tqdm


def set_seed(seed: int):
    random.seed(seed)

    np.random.seed(seed)

    torch.manual_seed(seed)

    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)

        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False


# Example usage
seed_value = 42
set_seed(seed_value)


# ------------------------------------------------------------------
# Contrastive Loss (InfoNCE)
# ------------------------------------------------------------------
def info_nce_loss(embeddings, temperature=0.07):
    """
    embeddings: shape (2N, e)
      The first N are "view 1" of volumes,
      the next N are "view 2" of volumes.
    temperature: float
    Returns:  loss
    """
    embeddings = F.normalize(embeddings, dim=1)  # (2N, e)
    N = embeddings.size(0) // 2

    emb_view1 = embeddings[:N, :]  # shape (N, e)
    emb_view2 = embeddings[N:, :]  # shape (N, e)

    sim_matrix = torch.mm(emb_view1, emb_view2.t())  # (N, N)

    # Convert sim to logits
    logits = sim_matrix / temperature

    target = torch.arange(N, device=logits.device)  # (N,)

    loss = F.cross_entropy(logits, target)
    return loss


class VolumeAggregator(nn.Module):
    def __init__(self, embed_dim=768, volume_dim=256):
        super().__init__()
        self.cls_token = nn.Parameter(torch.zeros(1, 1, embed_dim))
        encoder_layer = nn.TransformerEncoderLayer(
            d_model=embed_dim, nhead=8, dim_feedforward=1024
        )
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=2)

        self.volume_proj = nn.Linear(embed_dim, volume_dim)

    def forward(self, slice_embeddings):
        """
        slice_embeddings: (S, E) 

        Returns: A single volume embedding of shape (volume_dim,).
        """
        S = slice_embeddings.size(0)

        cls_token = self.cls_token.repeat(1, 1, 1)  # shape (1, 1, E)
        cls_token = cls_token.squeeze(0)  # shape (1, E)

        tokens = torch.cat([cls_token, slice_embeddings], dim=0)  # (S+1, E)

        tokens = tokens.unsqueeze(1)  # shape (S+1, 1, E)

        transformed = self.transformer_encoder(tokens)  # (S+1, 1, E)

        cls_output = transformed[0]  # shape (1, E)

        volume_embedding = self.volume_proj(cls_output)  # (1, volume_dim)
        volume_embedding = volume_embedding.squeeze(0)  # (volume_dim,)

        return volume_embedding


# ------------------------------------------------------------------
# Training Loop
# ------------------------------------------------------------------
class MultiViewContrastiveTrainer:
    def __init__(
        self, encoder_2d, aggregator_3d, device="cuda", lr=1e-4, temperature=0.07
    ):

        self.encoder_2d = encoder_2d.to(device)
        self.aggregator_3d = aggregator_3d.to(device)
        self.device = device
        self.temperature = temperature

        for param in self.encoder_2d.parameters():
            param.requires_grad = False  # freeze

        self.optimizer = torch.optim.Adam(
            filter(lambda p: p.requires_grad, aggregator_3d.parameters()), lr=lr
        )

    def forward_volume_view(self, volume_view):
        """
        volume_view: a tensor of slices for 1 volume
                     shape ~ (S, 1, H, W) for S slices.
        Returns a single embedding (volume_dim,).
        """
        slice_embeddings = []
        for slice_image in volume_view:  # shape (1, H, W)
            slice_tensor = slice_image.unsqueeze(0).to(self.device)  # (B=1, 1, H, W)
            z_2d = self.encoder_2d(slice_tensor)  # (1, E)
            slice_embeddings.append(z_2d.squeeze(0))  # (E,)

        slice_embeddings = torch.stack(slice_embeddings, dim=0)  # (S, E)
        v_emb = self.aggregator_3d(slice_embeddings)  # (volume_dim,)
        return v_emb

    def train_step(self, batch_of_volumes_view1, batch_of_volumes_view2):
        """
        batch_of_volumes: a list of volumes.
        """
        self.optimizer.zero_grad()
        volume_embeddings = []

        for viewA, viewB in zip(batch_of_volumes_view1, batch_of_volumes_view2):

            embA = self.forward_volume_view(viewA)
            embB = self.forward_volume_view(viewB)

            volume_embeddings.append(embA)
            volume_embeddings.append(embB)

        volume_embeddings = torch.stack(volume_embeddings, dim=0)

        loss = info_nce_loss(volume_embeddings, temperature=self.temperature)
        loss.backward()
        self.optimizer.step()

        return loss.item()


class MultiViewNiiDataset(Dataset):
    def __init__(
        self,
        folder_path,
        transform=None,
        num_slices=9,
        planes=["axial", "coronal"],
        slice_selector="normal",
        margin_between_slices=5,
    ):
        """
        Args:
            folder_path (str): Path to the folder containing .nii.gz files.
            num_slices (int): Number of slices to extract from each scan.
            plane (str): Plane of slices to extract ('axial', 'coronal', or 'sagittal').
            slice_selector (str): Method to select slices ('normal' or 'middle').
        """
        self.folder_path = folder_path
        self.file_names = [f for f in os.listdir(folder_path) if f.endswith(".nii.gz")]
        self.num_slices = num_slices
        self.planes = planes
        self.slice_selector = slice_selector
        self.margin_between_slices = margin_between_slices
        self.transform = transform

    def __len__(self):
        return len(self.file_names)

    def __getitem__(self, idx):
        file_name = self.file_names[idx]
        file_path = os.path.join(self.folder_path, file_name)

        nii_image = nib.load(file_path)
        image_data = nii_image.get_fdata()

        view1, view2 = None, None
        if len(self.planes) >= 1:
            view1_slices = self._get_slices(image_data, self.planes[0])
            view1 = torch.stack(
                [
                    self.transform(Image.fromarray(slice_.astype(np.uint8)))
                    for slice_ in view1_slices
                ]
            )
        if len(self.planes) > 1:
            view2_slices = self._get_slices(image_data, self.planes[1])
            view2 = torch.stack(
                [
                    self.transform(Image.fromarray(slice_.astype(np.uint8)))
                    for slice_ in view2_slices
                ]
            )

        return view1, view2  # Return two views as specified by planes

    def _get_slices(self, data, plane):
        """
        Extract slices from the 3D data based on the specified plane and slice selector.

        Args:
            data (np.ndarray): 3D numpy array of the scan data (shape: X, Y, Z).

        Returns:
            list: List of 2D slices as numpy arrays.
        """
        if plane not in ["axial", "coronal", "sagittal"]:
            raise ValueError("Plane must be one of 'axial', 'coronal', or 'sagittal'.")

        if self.slice_selector not in ["normal", "middle"]:
            raise ValueError(
                "Slice method is not define, it must be one of 'normal' or 'middle'"
            )

        if plane == "axial":
            z_dim = data.shape[2]
            if self.slice_selector == "normal":
                indices = np.linspace(0, z_dim - 1, self.num_slices, dtype=int)
            elif self.slice_selector == "middle":
                mid_index = z_dim // 2
                margin = self.margin_between_slices * self.num_slices

                start_index = mid_index - margin
                end_index = mid_index + margin

                start_index = max(0, start_index)
                end_index = min(z_dim - 1, end_index)

                indices = np.linspace(
                    start_index, end_index, self.num_slices, dtype=int
                )
            slices = [data[:, :, i] for i in indices]

        elif plane == "coronal":
            y_dim = data.shape[1]
            if self.slice_selector == "normal":
                indices = np.linspace(0, y_dim - 1, self.num_slices, dtype=int)
            elif self.slice_selector == "middle":
                mid_index = y_dim // 2
                margin = self.margin_between_slices * self.num_slices

                start_index = mid_index - margin
                end_index = mid_index + margin

                start_index = max(0, start_index)
                end_index = min(y_dim - 1, end_index)

                indices = np.linspace(
                    start_index, end_index, self.num_slices, dtype=int
                )
            slices = [data[:, i, :] for i in indices]

        elif plane == "sagittal":
            x_dim = data.shape[0]
            if self.slice_selector == "normal":
                indices = np.linspace(0, x_dim - 1, self.num_slices, dtype=int)
            elif self.slice_selector == "middle":
                mid_index = x_dim // 2
                margin = self.margin_between_slices * self.num_slices

                start_index = mid_index - margin
                end_index = mid_index + margin

                start_index = max(0, start_index)
                end_index = min(x_dim - 1, end_index)

                indices = np.linspace(
                    start_index, end_index, self.num_slices, dtype=int
                )
            slices = [data[i, :, :] for i in indices]

        return slices

```


and to run the code follow:

```python
if __name__ == "__main__":
    SSL_WEIGHT_PATH = "PATH_TO_2D_DINO_BACKBONE_WEIGHTS"
    DATASET_FOLDER = "PATH_TO_3D_CT_DATA"
    OUTPUT_PATH = "OUTPUT_PATH"

    backbone = torch.hub.load(
        "facebookresearch/dino:main", "dino_vitb16", pretrained=False
    )
    backbone.load_state_dict(torch.load(SSL_WEIGHT_PATH))

    aggregator_3d = VolumeAggregator(embed_dim=768, volume_dim=256)

    trainer = MultiViewContrastiveTrainer(
        encoder_2d=backbone,
        aggregator_3d=aggregator_3d,
        device="cuda",  # or 'cuda'
        lr=1e-4,
        temperature=0.1,
    )

    t = T.Compose(
        [
            T.Grayscale(num_output_channels=3),
            T.Resize(224, interpolation=T.InterpolationMode.BICUBIC),
            T.CenterCrop(224),
            T.ToTensor(),
        ]
    )

    dataset = MultiViewNiiDataset(
        folder_path=DATASET_FOLDER,
        num_slices=32,
        planes=["axial", "sagittal"],
        slice_selector="middle",
        margin_between_slices=4,
        transform=t,
    )

    dl = DataLoader(dataset=dataset, batch_size=32, shuffle=True)
    for epoch in range(100):
        total_loss = 0
        num_batches = len(dl)
        with tqdm(dl, desc=f"Epoch {epoch+1}", leave=False) as pbar:
            for i, (volumesA, volumesB) in enumerate(pbar):
                loss_val = trainer.train_step(volumesA, volumesB)
                total_loss += loss_val

                pbar.set_postfix({"Loss": f"{loss_val:.4f}"})

        avg_loss = total_loss / num_batches
        print(f"Epoch {epoch+1} | Average Loss: {avg_loss:.4f}")

        torch.save(
            trainer.aggregator_3d.state_dict(),
            os.path.join(OUTPUT_PATH, "projector.pth"),
        )
```

**Comment by Zhong1015** [comment]
Thank you for your reply. In fact, I can get a not bad result if i do not use the aggregator pretraining. 
As you said in the ablation study, I carefully read the ablation study part. 
I would like to confirm a few facts:
① If only SAMF is used, the Aggregator does not need to be pre-trained; however, if Ao2D is used, the Aggregator must be pre-trained. As shown in Table 3, when Ao2D is used without pre-training the Aggregator, the performance is very poor.

② Regarding the last row of Table 3—that is, the setting where Ao2D and SSL warm-up are used while both the visual encoder and the Aggregator are frozen—I would like to confirm whether BLEU-4 is actually higher than BLEU-1. It is clear that the performance under this setting is not good, but the table indicates that BLEU-4 can reach 0.4+. I would like to confirm whether this is correct, as I am somewhat confused.

③ I can understand the SAMF in the Table1 correspond to the Table3 SAMF. But I could not find the result in the Table 3 which is corresponding to the Table1(SAMF+Ao2d)

Thank you again for your help and time. I am looking forward to your response !

**Comment by abd-hsn** [comment]
I think there may be a misunderstanding regarding Table 3 and Table 1. To clarify:

**Table 1** reports results for both **SAMF** and **SAMF+Ao2D** *without* any warmup/pretraining stage and *without* freezing any components. **Table 3**, on the other hand, shows model performance when we **either freeze a component** or **apply a warmup stage**.

Now, addressing your points:

1. As stated in the paper, based on **Table 1** and **Table 3**, the SSL warmup does **not** improve model performance.
   For the **Ao2D** setup specifically:

   * The average of the *four scores **before*** pretraining the aggregator (7th row of Table 1) is **0.384**. While the average *after* pretraining the aggregator (6th row of Table 3) drops to **0.338**. I believe by “_poor performance in Table 3_” you are referring to the **5th row of Table 3**, where the **ViT is frozen**. We explicitly explain the reason for this behavior in the paper:
> This sensitivity can be attributed to Ao2D’s inherent reliance on the 2D embedding space derived from the image encoder for pattern recognition. Its architectural design, which focuses on attention over two dimensional features, makes it dependent on maintaining an adaptable 2D encoder.

2. Yes, the results are correct, and you can evaluate them using `eval_caption.py`.
   Regarding **BLEU-4 being higher than BLEU-1**, we were also surprised and do not have a definitive explanation for this behavior in this specific scenario.

3. I may be misunderstanding the question, but as mentioned earlier, **Table 3 differs from Table 1** in that it involves **freezing model components**, whereas Table 1 does not.

**Comment by Zhong1015** [comment]
Thank you for your response. After reading your response, I consider that the SSL warm up have little performance, so I can do all the experiments without the SSL warm up.
